cip-ledger-state.md

```diff
@@ -0,0 +1,111 @@
+configuration) that machine has two basically identical copies of ledger state. Ledger state is a
+*HUGE* data structure and the mainnet version currently consumes over 10 Gigabytes of memory.
+Furthermore, maintaining ledger state duplicates code functionality that is in `ouroboros-consensus`
+and the maintenance of ledger state has been the cause of about 90% of the bugs in `db-sync` over
+the last two years. The maintenance of ledger state also makes updating, running and maintaining
+`db-sync` by operators more difficult than it should be. Finally, if `db-sync` did not have to
+maintain ledger state, the size of the `db-sync` code base would probably decrease by about 50% and
+the bits removed are some of the most complicated parts.
+
+
+## Specification
+
+The proposed solution is an enhanced local chain sync protocol that would only be served over a
+local domain socket. The enhanced chain chain sync protocol would include the existing block chain
+data as well as events containing the ledger data that db-sync needs. This enhanced local chain
+sync protocol is useful for many applications other than just db-sync.
+
+Smart contract developers would like an application that turns block chain and ledger state changes
+into an event stream. With this enhanced local chain sync protocol, generating an easily consumable
+event stream simply requires a conversion of the binary enhanced local chain sync protocol into
+JSON.
+
+This enhanced local chain sync protocol is basically the data that would be provided by the
+proposed Scientia program (which as far as I am aware has been dropped).
+
+The ledger state data that would provided over the extended chain sync protocol is limited to:
+
+* Per epoch stake distribution (tickle fed in a deterministic manner).
+* Per epoch rewards (tickle fed in a deterministic manner).
+* Per epoch protocol parameters (tiny so provide in a single event)
+* Per epoch reaped pool list (single event).
+* Per epoch MIR distribution (single event).
+* Per epoch pool deposit refunds (single event).
+
+The small bits of data above are sent as single per epoch events. Large bits of data like the epoch
+stake distribution map (which can have millions of entries) are tickle fed as they are calculated
+by the ledger. Its deterministic in that give the same ledger state LS, with the same block B, the
+events generated will always be identical. This is so that if the consumer is stopped and restarted
+and needs to rollback a block or two, the replay will be identical.
+
+
+## Rationale
+
+THe recommended configuration for `db-sync` is to run it on the same machine as the `node`.
+Currently this means that there are two copies of the *HUGE* ledger state data structure (each being
+at least 10G in size) on the machine. In addition, `db-sync` and other applications only need about
+1% of that data. The rest is
+
+
+## Test Cases
+
+
+
+## Implementations
```

Comment:
A possible implementation first step would be to write a prototype funnelling existing events (or only one of them for simplicity's sake because I assume there would be need to implement serialisation) through the consensus code down to the network stack, in order to properly understand the impact of this change.